jetty-server/src/main/java/org/eclipse/jetty/server/AbstractLockedHttpInput.java

```diff
@@ -0,0 +1,236 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.io.IOException;
+import javax.servlet.ReadListener;
+
+import org.eclipse.jetty.util.thread.AutoLock;
+
+public abstract class AbstractLockedHttpInput extends AbstractHttpInput
```

Comment:
Rather than have both `AbstractLockedHttpInput` and `AbstractHttpInput`, I think `AbstractHttpInput` should just be lockable but initialised with a NOOP AutoLock that will be removed by the JIT

Reply:
That sounds like a reasonable idea. This code is still in motion, so I haven't given this particular subject too much thought so far. But I'd prefer to wait until all the functionality we want to be fully working before working on that.

Reply:
ack!    Although I think the NOOP lock idea will not work because `AbstractHttpInput` still needs a real lock in some situations....
So consider this comment to be that ultimately all implementations should have the same locking behaviour and if not then we really need to be clear why not.

Reply:
Agreed. Now that all three implementations are working, we need to investigate their differences to make educated decisions on what to do about them.